usr/lib/pkcs11/common/sign_mgr.c

```diff
@@ -60,21 +60,36 @@ CK_RV sign_mgr_init(STDLL_TokData_t *tokdata,
         else
             return rc;
     }
-    // is key allowed to generate signatures?
-    //
-    rc = template_attribute_find(key_obj->template, CKA_SIGN, &attr);
-    if (rc == FALSE) {
-        TRACE_ERROR("Could not find CKA_SIGN for the key.\n");
-        return CKR_KEY_FUNCTION_NOT_PERMITTED;
+
+    if (recover_mode) {
+        // is key allowed to generate signatures where the data can be
+        // recovered from the signature?
+        rc = template_attribute_find(key_obj->template, CKA_SIGN_RECOVER,
+                                     &attr);
+        if (rc == FALSE) {
+            TRACE_ERROR("Could not find CKA_SIGN_RECOVER for the key.\n");
+            return CKR_KEY_FUNCTION_NOT_PERMITTED;
+        }
+        flag = *(CK_BBOOL *) attr->pValue;
+        if (flag != TRUE) {
+            TRACE_ERROR("%s\n", ock_err(ERR_KEY_FUNCTION_NOT_PERMITTED));
+            return CKR_KEY_FUNCTION_NOT_PERMITTED;
+        }
```

Comment:
Code for setting and checking flag is doubled in if-else. Could you move it below if-else ?

Reply:
fixed